File: week-2-arrays/problem-set/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

string get_text(void);
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
void print_grade(int words, int letters, int sentences);

int main(void)
{
    // Get input string from user
    string text = get_text();

    // Count number of letters
    int letters = count_letters(text);

    // Count number of words
    int words = count_words(text);

    // Count number of sentences
    int sentences = count_sentences(text);

    // Print the result
    print_grade(letters, words, sentences);
}

// Prompt the user
string get_text(void)
{
    string s;
    do
    {
        s = get_string("Text: ");
    }
    while (s[0] < 1);
    return s;
}

// Count the number of letters in the given text
int count_letters(string text)
{
    int letters = 0;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }

    return letters;
}

// Count the number of words in the given text
int count_words(string text)
{
    int words = 0;
    bool in_word = false;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (!isspace(text[i]) && !in_word)
        {
            words++;
            in_word = true;
        }
        else if (isspace(text[i]))
        {
            in_word = false;
        }
    }

    return words;
}

// Count the number of sentences in the given text
int count_sentences(string text)
{
    int sentences = 0;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (!ispunct(text[i]) && (text[i + 1] == '.' || text[i + 1] == '!' || text[i + 1] == '?'))
        {
            sentences++;
        }
    }

    return sentences;
}

// Print the grade based on Coleman-Liau index formula
void print_grade(int letters, int words, int sentences)
{
    float L = (float) letters / words * 100;
    float S = (float) sentences / words * 100;
    float grade = 0.0588 * L - 0.296 * S - 15.8;

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", (int) round(grade));
    }
}

File: week-6-python/problem-set/sentimental-readability/readability.py
# TODO
def main():
    text = get_text()

    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    print(calculate_grades(letters, words, sentences))


def get_text():
    while True:
        # Get user input for the text
        text = input("Text: ")
        # Check if the text is not an empty string
        if len(text) > 0:
            return text


def count_letters(text):
    # Initialize a variable to store the count of letters
    letters = 0

    # Iterate each char in the text
    for char in text:
        # Check if char is letter or not
        if char.isalpha():
            # Increment letters
            letters += 1

    return letters


def count_words(text):
    # Initialize variable to store values
    words = 0
    in_word = False

    # Iterate each char in text
    for char in text:
        # Check if char is in word or not
        if not char.isspace() and not in_word:
            # Increment words
            words += 1
            in_word = True
        elif char.isspace():
            in_word = False

    return words


def count_sentences(text):
    # Initialize a variable to store the count of sentence
    sentences = 0

    # Iterate each char in text
    for i in range(len(text) - 1):
        # Check if current char is alnum and next char is ., !, or ?
        if text[i].isalnum() and text[i + 1] in [".", "!", "?"]:
            # Increment sentences
            sentences += 1

    return sentences


def calculate_grades(letters, words, sentences):
    # Calculate the percentage of letters and sentences relative to the number of words
    L = float(letters / words * 100)
    S = float(sentences / words * 100)

    # Calculate the Coleman-Liau index
    grade = 0.0588 * L - 0.296 * S - 15.8

    # Determine the grade level based on the calculated index and then return
    if grade < 1:
        return "Before Grade 1"
    elif grade >= 16:
        return "Grade 16+"
    else:
        return f"Grade {round(grade)}"


if __name__ == "__main__":
    main()
